### ISSUE TITLE
Flink S3Fileio incorrect commit

### ISSUE THREAD

**Issue by krisdas** [apache/iceberg#4168]
Flink version : 1.13.2
Iceberg version : 0.12
We were trying S3Fileio instead of default hadoop fileio, to ingest data into iceberg on S3.

While commit/saving checkpoint, we see below call stack while uploading data file. So looks datafile (for example : s3/bucket/.../iceberg_db.db/iceberg_table/data/file.parquet) upload fails.

```
exception: { [-] 
      exception_class:  java.util.concurrent.CompletionException 
      exception_message:  java.io.UncheckedIOException: java.nio.file.NoSuchFileException: /tmp/s3fileio-224113294985193160.tmp 
      stacktrace:  java.util.concurrent.CompletionException: java.io.UncheckedIOException: java.nio.file.NoSuchFileException: /tmp/s3fileio-224113294985193160.tmp
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1702)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.io.UncheckedIOException: java.nio.file.NoSuchFileException: /tmp/s3fileio-224113294985193160.tmp
	at software.amazon.awssdk.utils.FunctionalUtils.asRuntimeException(FunctionalUtils.java:180)
	at software.amazon.awssdk.utils.FunctionalUtils.lambda$safeSupplier$4(FunctionalUtils.java:110)
	at software.amazon.awssdk.utils.FunctionalUtils.invokeSafely(FunctionalUtils.java:136)
	at software.amazon.awssdk.core.sync.RequestBody.fromFile(RequestBody.java:88)
	at software.amazon.awssdk.core.sync.RequestBody.fromFile(RequestBody.java:99)
	at org.apache.iceberg.aws.s3.S3OutputStream.lambda$uploadParts$1(S3OutputStream.java:237)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	... 3 more
Caused by: java.nio.file.NoSuchFileException: /tmp/s3fileio-224113294985193160.tmp
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:116)
	at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
	at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:149)
	at java.base/sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:99)
	at java.base/java.nio.file.Files.readAttributes(Files.java:1764)
	at java.base/java.nio.file.Files.size(Files.java:2381)
	at software.amazon.awssdk.core.sync.RequestBody.lambda$fromFile$0(RequestBody.java:88)
	at software.amazon.awssdk.utils.FunctionalUtils.lambda$safeSupplier$4(FunctionalUtils.java:108)
	... 8 more
```

But on the iceberg manifest file, that data file (s3/bucket/.../iceberg_db.db/iceberg_table/data/file.parquet) is still referenced, as result, iceberg thinks that data file is valid and exists.

Now running another query, which needs to use the above data file, throws exception.

Has anybody seen this scenario? We are trying to re-produce the scenario for further investigation.

Code link : https://github.com/apache/iceberg/blob/master/aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java#L291

**Comment by dramaticlly** [comment]
So essentially what we observed is that the Flink S3 FileIO failed to upload the data files due to file not exist, but the missing data file is incorrectly tracked in iceberg manifest. So the subsequent query against the partition will fail as the claimed data file cannot be found.

Our iceberg was setup on AWS S3 with versioned bucket and we can confirm that the 
- data file never get uploaded, no version exists of given path
- the data file is being tracked in iceberg and we do see such non-exist data file in iceberg metadata query

This result in some nasty behavior where we need to reconcile the manifest state based on what exists in underlying file system, which is not possible. So we had to drop the partition as temporary work around. But I think it would be helpful to understand 

why do we run into the issue where iceberg commits before s3 confirm the data file is uploaded?

In the meantime, we are trying to see if we can have a simple repro of the issue

CC @szehon-ho

**Comment by jackye1995** [comment]
cc @danielcweeks do you see similar issues in your environment for S3FileIO? Sounds like we might be swallowing some exceptions in the progress multi-part upload logic

**Comment by szehon-ho** [comment]
FYI  this may be related to https://github.com/apache/iceberg/pull/4673, at the time if there's OOM but commit successful, cleanup may be done erroneously to remove valid files to corrupt the table.  It was done in Spark, not sure if on Flink it's the same pattern.

**Comment by danielcweeks** [comment]
From the stack trace, it looks like this is a failure during part upload, but that should result in an abort of the upload request.  It may be the case that the abort is happening, but it's possible that we're not throwing in the failed part upload path here: https://github.com/apache/iceberg/blob/master/aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java#L317

**Comment by danielcweeks** [comment]
I did a little investigation and this should actually be handled correctly.  The completable future is [joined on](https://github.com/apache/iceberg/blob/master/aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java#L331) which will propagate the exception up prior to closing the stream.  I'm not clear how a task could have completed successfully without hitting this error unless it is an issue higher up in Flink.

**Comment by stevenzwu** [comment]
@szehon-ho Flink doesn't do any abort cleanup that delete data files when commit failed. 

From the stack trace, it seems that the staging/part file was not found in local disk (tmp dir). How could that happen? disk full? but even if disk is full, we also need a bug somewhere that prevents the error from bubbling up.

**Comment by krisdas** [comment]
@stevenzwu as far i remember, we searched for any logs saying, disk is full or the staging file was not created due to x reason. but we couldn't find anything on that line.

**Comment by stevenzwu** [comment]
@krisdas do you have the complete stack trace?

**Comment by krisdas** [comment]
Unfortunately that's all we have now.

**Comment by dramaticlly** [comment]
Just want to share another incidence of this happening in the our production Flink application last week. This time we added extra instrumentation and collected more data points, but we still failed to get a reproduce of the issue

## Setup 
Iceberg 0.13.0 and Flink 13 with S3 fileIO turned on (default config for MPU size and threshold)

## Symptom
- Iceberg committed a broken snapshot `3619033567453770968` which track a non-exist data file `00008-0-a771cfe9-b71d-4e84-b784-e3b270b0ff80-00514.parquet` at  2022-07-10 06:15:00.666 PST
- Subsequent read of given partition failed with `FileNotFoundException` because process engine cannot find the problematic data files on S3

## Mitigation
- Use scripts from iceberg library to "delete" the dataFiles, something like
```scala
val table = getIcebergTable(s"$db_name.$tbl_name")
val df = spark.sql(s"select * from iceberg.$db_name.$tbl_name.files")
val dataf = df.filter(df("file_path") === dataFileOfInterest).persist

val origDataFileSize=dataf.select("file_size_in_bytes").map(r => r.getLong(0)).collect().head
val origDataFileRecordCount=dataf.select("record_count").map(r => r.getLong(0)).collect().head
val dt = DataFiles.builder(table.spec)
.withPath(dataFileOfInterest)
.withFileSizeInBytes(origDataFileSize)
.withPartitionPath(origPartitionPath)
.withRecordCount(origDataFileRecordCount)
.build()

val t = table.newTransaction
t.newOverwrite().deleteFile(dt).commit()
t.commitTransaction();
```
- However some of data consumer are using [Spark incremental read](https://iceberg.apache.org/docs/latest/spark-queries/#incremental-read) to get list of appended data files in between snapshots and mitigation above can only help with read from latest snapshot, so there is no easy way for us to fix snapshot `3619033567453770968` but skip processing of this corrupted snapshot, which result in some data loss

## Some investigation
- We spent some time to collect the requests from AWS side and realized the S3 multipart upload get aborted at 2022-07-10 06:07:01.619 PST, or 8 minutes before snapshot was committed
- The multipart upload abortion was caused by transient error in 1 of the upload-parts like `software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Error writing to server` and all subsequent parts upload failed as expected.
- Initially we are thinking maybe there's some exception gets swallowed so caller of https://github.com/apache/iceberg/blob/master/aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java might not realize the problem of writing this parquet data files to S3. But we also spent some time in this AWS integration tests https://github.com/apache/iceberg/blob/master/aws/src/integration/java/org/apache/iceberg/aws/s3/TestS3MultipartUpload.java and looks like S3OutputStream behave as expected -> If there's exception in threads to upload part, `java.util.concurrent.CompletionException` will bubble up as expected on `close()`

## Questions
- Correct expectation of using S3OutputStream? Is it safe to assume that iceberg is ready to commit if all writers can write its data and metadata files to fileIO and writers are close without exception?
- We are curious about call path on how S3OutputStream is wired to IcebergStreamWriter ? (assume this is where commit of list of data files happens), there's a lot of complexity in between and hard to identify the exact caller of S3OutputStream
- What else does community suggest to help root cause the actual problem and prevent it from happening again?

CC @rdblue @danielcweeks @szehon-ho @jackye1995 @singhpk234 @stevenzwu

**Comment by rdblue** [comment]
> Is it safe to assume that iceberg is ready to commit if all writers can write its data and metadata files to fileIO and writers are close without exception?

Yes. That's why I'm surprised that you ended up with corrupt metadata. The complete failure should have caused a task failure that triggered a retry, rather than producing data. My guess is that there is an issue with error handling in the Flink writer that we should look into.

**Comment by dramaticlly** [comment]
> Was the table partitioned or not? That will tell us the internal writer that was used.

Yes the table is partitioned by date and hour

> Since the multipart upload failed well before the snapshot was committed, my guess is that it happened when rolling a data file in the RollingFileWriter. I don't see a way that would get swallowed, but it is a good starting point to check into more.

Thanks for the pointer, I will dive into this and check


> Do you have any stack trace from the logs where the failure happened?

So here's the log from 1st failed UploadPart call
```log
   message: Failed to upload part: UploadPartRequest(Bucket=redactedBucket, ContentLength=33554432, Key=hive/warehouse/redact_prod.db/redact_event_f/data/date=2022-07-10/hour=13/00008-0-a771cfe9-b71d-4e84-b784-e3b270b0ff80-00514.parquet, PartNumber=5, UploadId=7.mWLRQMPhA0n9LjxP1E2oe6yEwXfdrP2k66qzDNzxgg.0usbpYhOGOYFvuNAcauzhgxDZqFnLxKFE68L7N5uNtrxD6Xxww2EFjJ_oiDilnXFq6YR02_SbNPmLpDhd8sxNJXQHwmhJrj.Wd7_cOOjXuicNzH4B56O5XhnM9gwog-)
```

and corresponding stacktrace
```
exception: { 
     exception_class: java.util.concurrent.CompletionException
     exception_message: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Error writing to server
     stacktrace: java.util.concurrent.CompletionException: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Error writing to server
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1702)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: software.amazon.awssdk.core.exception.SdkClientException: Unable to execute HTTP request: Error writing to server
	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:98)
	at software.amazon.awssdk.core.exception.SdkClientException.create(SdkClientException.java:43)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.utils.RetryableStageHelper.setLastException(RetryableStageHelper.java:205)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:66)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:34)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:56)
	at software.amazon.awssdk.core.internal.http.StreamManagingStage.execute(StreamManagingStage.java:36)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.executeWithTimer(ApiCallTimeoutTrackingStage.java:80)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:60)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallTimeoutTrackingStage.execute(ApiCallTimeoutTrackingStage.java:42)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:48)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallMetricCollectionStage.execute(ApiCallMetricCollectionStage.java:31)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:37)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ExecutionFailureExceptionReportingStage.execute(ExecutionFailureExceptionReportingStage.java:26)
	at software.amazon.awssdk.core.internal.http.AmazonSyncHttpClient$RequestExecutionBuilderImpl.execute(AmazonSyncHttpClient.java:193)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.invoke(BaseSyncClientHandler.java:133)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.doExecute(BaseSyncClientHandler.java:159)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.lambda$execute$1(BaseSyncClientHandler.java:112)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.measureApiCallSuccess(BaseSyncClientHandler.java:167)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:94)
	at software.amazon.awssdk.core.client.handler.SdkSyncClientHandler.execute(SdkSyncClientHandler.java:45)
	at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:55)
	at software.amazon.awssdk.services.s3.DefaultS3Client.uploadPart(DefaultS3Client.java:9609)
	at org.apache.iceberg.aws.s3.S3OutputStream.lambda$uploadParts$2(S3OutputStream.java:308)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
	... 3 more
Caused by: java.io.IOException: Error writing to server
	at java.base/sun.net.www.protocol.http.HttpURLConnection.writeRequests(HttpURLConnection.java:718)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.writeRequests(HttpURLConnection.java:730)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1613)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1520)
	at java.base/java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:527)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:334)
	at software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient$RequestCallable.call(UrlConnectionHttpClient.java:210)
	at software.amazon.awssdk.http.urlconnection.UrlConnectionHttpClient$RequestCallable.call(UrlConnectionHttpClient.java:193)
	at software.amazon.awssdk.core.internal.util.MetricUtils.measureDurationUnsafe(MetricUtils.java:64)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeHttpRequestStage.executeHttpRequest(MakeHttpRequestStage.java:76)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeHttpRequestStage.execute(MakeHttpRequestStage.java:55)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.MakeHttpRequestStage.execute(MakeHttpRequestStage.java:39)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.RequestPipelineBuilder$ComposingRequestPipelineStage.execute(RequestPipelineBuilder.java:206)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:73)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptTimeoutTrackingStage.execute(ApiCallAttemptTimeoutTrackingStage.java:42)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:77)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.TimeoutExceptionHandlingStage.execute(TimeoutExceptionHandlingStage.java:39)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptMetricCollectionStage.execute(ApiCallAttemptMetricCollectionStage.java:50)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.ApiCallAttemptMetricCollectionStage.execute(ApiCallAttemptMetricCollectionStage.java:36)
	at software.amazon.awssdk.core.internal.http.pipeline.stages.RetryableStage.execute(RetryableStage.java:64)
	... 27 more

   }
```

**Comment by dramaticlly** [comment]
>Looks like one problem is that the exception is [logged but not rethrown in the async future](https://github.com/apache/iceberg/blob/master/aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java#L309-L318). Because no exception is thrown, the call to CompleteableFuture::join in completeMultipartUpload succeeds and should return null for that part.

Yes @rdblue and I think we also observed that, and align with what @danielcweeks suggested earlier in https://github.com/apache/iceberg/issues/4168#issuecomment-1123000990. I think rethrow after logging is desired, but subsequent multi-part uploads will also fail due to previous abort. We also attempted to reproduce abort MPU in integration tests , verified we can assert thrown exceptions on close.

> Either way, it looks like some exception should cause close to fail, but that is apparently ignored since the writer creates metadata for the file.

yep I think if there's any suggestion/pointers can help us understand why this is happening, it would be of tremendous help. Today we are seeing 3 more occurrence of this happening, some is related to flink TM out of memory/heap issue and somes are not, so have to disable S3FileIO for flink before this is root caused and prevention is in-place. On the other side, we have turned S3FileIO on for Spark for months now but never see any issue of similar kind.

**Comment by abmo-x** [comment]
Wanted to also add that we are seeing similar behavior for file which BaseTaskWriter logs as successfully completed and I don't see any errors for s3 upload in the logs around that time frame. 

But the data file does not exist on S3, but the below log is logged which then adds the data file to the manifest.

```
logger_name:  org.apache.iceberg.io.BaseTaskWriter 
Complete data file GenericDataFile{content=data, file_path=PATH_TO_FILE_THAT_DOESN'T EXIST file_format=PARQUET, spec_id=0, partition=PartitionData{date=2022-07-15, hour=14}
```

Our S3 bucket is versioned, no versions exists in s3 for that file.

**Comment by rdblue** [comment]
The fix for this was released in 0.14.1
